src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -158,91 +209,91 @@ internal ProjectRootElementCache(bool autoReloadFromDisk, bool loadProjectsReadO
             // Should already have been canonicalized
             ErrorUtilities.VerifyThrowInternalRooted(projectFile);
 
+            ProjectRootElement projectRootElement;
             lock (_locker)
             {
-                ProjectRootElement projectRootElement;
                 _weakCache.TryGetValue(projectFile, out projectRootElement);
 
+                if (projectRootElement != null)
+                {
+                    BoostEntryInStrongCache(projectRootElement);
+
+                    // An implicit load will never reset the explicit flag.
+                    if (isExplicitlyLoaded)
+                    {
+                        projectRootElement.MarkAsExplicitlyLoaded();
+                    }
+                }
+                else
+                {
+                    DebugTraceCache("Not found in cache: ", projectFile);
+                }
+
                 if (preserveFormatting != null && projectRootElement != null && projectRootElement.XmlDocument.PreserveWhitespace != preserveFormatting)
                 {
                     //  Cached project doesn't match preserveFormatting setting, so reload it
                     projectRootElement.Reload(true, preserveFormatting);
```

Comment:
We're still loading XML under the lock here. Intentional?

Reply:
Yes, intentional. 
This if section fixes a rare bug, and this code path uses it's own code for reloading. 
To refactor it such that we can divide the actual download and updating the element of the cache which has to be under the lock is not worth it, so we decided.